nifi-external/nifi-kafka-connect/nifi-kafka-connector/src/main/java/org/apache/nifi/kafka/connect/StatelessNiFiSinkTask.java

```diff
@@ -0,0 +1,329 @@
+            failurePortNames = Collections.emptySet();
+        } else {
+            failurePortNames = new HashSet<>();
+
+            final String[] names = failurePortList.split(",");
+            for (final String name : names) {
+                final String trimmed = name.trim();
+                failurePortNames.add(trimmed);
+            }
+        }
+
+        // Validate the failure ports
+        final Set<String> outputPortNames = dataflow.getOutputPortNames();
+        for (final String failurePortName : failurePortNames) {
+            if (!outputPortNames.contains(failurePortName)) {
+                throw new ConfigException("Dataflow was configured with a Failure Port of " + failurePortName
+                    + " but there is no Port with that name in the dataflow. Valid Port names are " + outputPortNames);
+            }
+        }
+
+        backgroundTriggerExecutor = Executors.newFixedThreadPool(1, r -> {
+            final Thread thread = Executors.defaultThreadFactory().newThread(r);
+            thread.setName("Execute dataflow " + dataflowName);
+            return thread;
+        });
+    }
+
+    @Override
+    public void put(final Collection<SinkRecord> records) {
+        if (!lastTriggerSuccessful) {
+            // When Kafka Connect calls #put, it expects the method to return quickly. If the dataflow should be triggered, it needs to be triggered
+            // in a background thread. When that happens, it's possible that the dataflow could fail, in which case we'd want to rollback and re-deliver
+            // the messages. Because the background thread cannot readily do that, it sets a flag, `lastTriggerSuccessful = false`. We check this here,
+            // so that if a background task failed, we can throw a RetriableException, resulting in the messages being requeued. Because of that, we ensure
+            // that any time that we set lastTriggerSuccessful, we also purge any data in the dataflow, so that it can be redelivered and retried.
+
+            lastTriggerSuccessful = true; // We don't want to throw a RetriableException again.
+            throw new RetriableException("Last attempt to trigger dataflow failed");
+        }
+
+        logger.debug("Enqueuing {} Kafka messages", records.size());
+
+        for (final SinkRecord record : records) {
+            final Map<String, String> attributes = createAttributes(record);
+            final byte[] contents = getContents(record.value());
+
+            queueSize = dataflow.enqueue(contents, attributes, inputPortName);
+        }
+
+        // If we haven't reached the preferred back size, return.
+        if (queueSize == null || queueSize.getObjectCount() < batchSize) {
+            return;
+        }
+        if (queueSize.getByteCount() < batchBytes) {
+            return;
+        }
+
+        logger.debug("Triggering dataflow in background thread");
+
+        backgroundTriggerExecutor.submit(this::triggerDataflow);
```

Comment:
As we discussed offline, would be preferable to do the trigger in flush, and only enqueue in put.

Reply:
Will do.